lib/galaxy/tools/provided_metadata.py

```diff
@@ -0,0 +1,185 @@
+import json
+import logging
+import os
+
+
+log = logging.getLogger(__name__)
+
+
+def parse_tool_provided_metadata(meta_file, provided_metadata_style=None, job_wrapper=None):
+    """Return a ToolProvidedMetadata object for specified file path.
+
+    If path is absent, return a NullToolProvidedMetadata. If type is None
+    attempt to guess tool provided metadata type.
+    """
+    if not os.path.exists(meta_file):
+        return NullToolProvidedMetadata()
+    if provided_metadata_style is None:
+        provided_metadata_style = _guess_tool_provided_metadata_style(meta_file)
+
+    assert provided_metadata_style in ["legacy", "default"]
+
+    if provided_metadata_style == "legacy":
+        return LegacyToolProvidedMetadata(meta_file, job_wrapper=job_wrapper)
+    elif provided_metadata_style == "default":
+        return ToolProvidedMetadata(meta_file)
+
+
+def _guess_tool_provided_metadata_style(path):
+    try:
+        with open(path, "r") as f:
+            metadata = json.load(f)
+        metadata_type = metadata.get("type", None)
+        return "legacy" if metadata_type in ["dataset", "new_primary_dataset"] else "default"
+    except ValueError:
+        # Either empty or multiple JSON lines, either way we can safely treat
+        # it as legacy style.
+        return "legacy"
+
+
+class ToolProvidedMetadata(object):
+
+    def get_new_datasets(self, output_name):
+        """Find new datasets for dataset discovery for specified output.
+
+        Return a list of such datasets.
+
+        Called only in the context of discovering datasets when
+        discover_via="tool_provided_metadata" is defined in the tool.
+        """
+        return []
+
+    def has_failed_outputs(self):
+        """Determine if generation of any of the outputs failed.
+        """
+        return False
+
+    def get_new_dataset_meta_by_basename(self, output_name, basename):
+        """For a discovered dataset, get the corresponding metadata entry.
+
+        The discovery may have been from explicit listing in this file (returned
+        from get_new_datasets) or via file regex, either way the basename of the
+        file is used to index the fetching of the metadata entry.
+        """
+        return {}
+
+    def get_unnamed_outputs(self):
+        """Return unnamed outputs dataset structured introduced for upload 2.0.
```

Comment:
Should this be 
```suggestion
        """Return unnamed outputs dataset introduced for upload 2.0.
```
?